runtime/compiler/optimizer/InterpreterEmulator.hpp

```diff
@@ -0,0 +1,253 @@
+   };
+
+/*
+ * \class KnownOperand
+ *
+ * \brief represent operands that can be reasoned about at compile time
+ */
+class KnownOperand : public Operand
+   {
+   public:
+      TR_ALLOC(TR_Memory::Inliner);
+      virtual bool isUnkownOperand(){ return false;}
+   };
+
+class IconstOperand : public KnownOperand
+   {
+   public:
+      TR_ALLOC(TR_Memory::Inliner);
+      IconstOperand (int x): intValue(x) { }
+      virtual IconstOperand *asIconst() { return this;}
+      int32_t intValue;
+   };
+
+class KnownObjOperand : public KnownOperand
+   {
+   public:
+      TR_ALLOC(TR_Memory::Inliner);
+      KnownObjOperand(TR::KnownObjectTable::Index koi):knownObjIndex(koi){ }
+      virtual KnownObjOperand *asKnownObject(){ return this;}
+      virtual TR::KnownObjectTable::Index getKnownObjectIndex(){ return knownObjIndex;}
+      TR::KnownObjectTable::Index knownObjIndex;
+   };
+
+/*
+ * \class KnownObjPairOperand
+ *
+ * \brief represents a pair of known objects
+ *
+ * \note This class is used to support mutable callsite because both the methodhandle object
+ *       and the mutable callsite needs to be tracked so that when creating \c TR_J9MutableCallSite
+ *       the mutable callsite object can be set for the callsite even though it's really the
+ *       methodhandle object that's on the operand stack.
+ *
+ * \see getReturnValueForInvokevirtual
+ * \see refineResolvedCalleeForInvokestatic
+ * \see visitInvokestatic
+ */
+class KnownObjPairOperand : public KnownOperand
+   {
+   public:
+      TR_ALLOC(TR_Memory::Inliner);
+      KnownObjPairOperand (TR::KnownObjectTable::Index koi1, TR::KnownObjectTable::Index koi2){ knownObjIndexPair = std::make_pair(koi1, koi2);}
+      virtual KnownObjPairOperand* asKnownObjectPair(){ return this;}
+      virtual TR::KnownObjectTable::Index getKnownObjectIndex(){ return knownObjIndexPair.first;}
+      std::pair<TR::KnownObjectTable::Index, TR::KnownObjectTable::Index> knownObjIndexPair;
+   };
+
+class InterpreterEmulator : public TR_ByteCodeIteratorWithState<TR_J9ByteCode, J9BCunknown, TR_J9ByteCodeIterator, Operand *>
+   {
+   typedef TR_ByteCodeIteratorWithState<TR_J9ByteCode, J9BCunknown, TR_J9ByteCodeIterator, Operand *> Base;
+
+   public:
+      InterpreterEmulator(
+            TR_CallTarget *calltarget,
+            TR::ResolvedMethodSymbol * methodSymbol,
+            TR_J9VMBase * fe,
+            TR::Compilation * comp,
+            TR_InlinerBase *inliner,
+            TR_EstimateCodeSize *ecs)
+         : Base(methodSymbol, comp),
+           _calltarget(calltarget),
+           _inliner(inliner),
+           _ecs(ecs),
+           _iteratorWithState(false)
+         {
+         TR_J9ByteCodeIterator::initialize(static_cast<TR_ResolvedJ9Method *>(methodSymbol->getResolvedMethod()), fe);
+         _flags = NULL;
+         _stacks = NULL;
+         }
+      TR_InlinerTracer *tracer() { return _inliner->tracer();}
```

Comment:
I feel something this complex probably warrants its own trace flag and trace levels...

Reply:
Modified accordingly to make InterpreterEmulator have it's own tracer.